contributors/design-proposals/pod-resource-management.md

```diff
@@ -1,114 +1,419 @@
+dynamically in response to the following events:
+
+1. kubelet startup/recovery
+1. prior to creation of the pod level cgroup
+1. after deletion of the pod level cgroup
+1. at periodic intervals to reach `experimental-qos-reserve-requests`
+heurisitcs that converge to a desired state.
+
+All writes to the QoS level cgroup sandboxes are protected via a
+common lock in the kubelet to ensure we do not have multiple concurrent
+writers to this tier in the hierarchy.
+
+### QoS level CPU allocation
+
+The `BestEffort` cgroup sandbox is statically configured as follows:
+
+```
+ROOT/BestEffort/cpu.shares = 2
+```
+
+This ensures that allocation of CPU time to pods in this QoS class
+is given the lowest priority.
+
+The `Burstable` cgroup sandbox CPU share allocation is dynamic based
+on the set of pods currently scheduled to the node.  
+
+```
+ROOT/Burstable/cpu.shares = max(sum(Burstable pods cpu requests, 2)
+```
+
+The Burstable cgroup sandbox is updated dynamically in the exit
+points described in the previous section.  Given the compressible
+nature of CPU, and the fact that cpu.shares are evaluated via relative
+priority, the risk of an update being incorrect is minimized as the `kubelet`
+converges to a desired state.  Failure to set `cpu.shares` at the QoS level
+cgroup would result in `500m` of cpu for a Guaranteed pod to have different
+meaning than `500m` of cpu for a Burstable pod in the current hierarchy.
+For this reason, we will always set `cpu.shares` for the QoS level sandboxes
+by default as part of roll-out for this feature.
+
+### QoS level memory allocation
+
+By default, no memory limits are applied to the BestEffort
+and Burstable QoS level cgroups unless a `--qos-reserve-requests` value
+is specified for memory.
+
+The heuristic that is applied is as follows for each QoS level sandbox:
+
+```
+ROOT/Burstable/memory.limit_in_bytes = 
+    Node.Allocatable - {(summation of memory requests of `Guaranteed` pods)*(reservePercent / 100)}
+ROOT/BestEffort/memory.limit_in_bytes = 
+    Node.Allocatable - {(summation of memory requests of all `Guaranteed` and `Burstable` pods)*(reservePercent / 100)}
+```
+
+A value of `--experimental-qos-reserve-requests=memory=100%` will cause the
+`kubelet` to adjust the Burstable and BestEffort cgroups from consuming memory
+that was requested by a higher QoS class. This increases the risk
+of inducing OOM on BestEffort workloads in favor of increasing memory
+resource guarantees for Burstable and Guaranteed workloads.  A value of 
+`--experimental-qos-reserve-requests=memory=0%` will allow a Burstable
+and BestEffort QoS sandbox to consume up to the full allocatable amount if
+available in favor of increasing memory resource guarantees for Guaranteed
+workloads.
+
+Since memory is an incompressible resource, it is possible that a QoS
+level cgroup sandbox may not be able to reduce memory usage below the
+value specified in the heuristic during pod admission and pod termination.
```

Comment:
We do use any `heuristic` do we? Limits are based on the policy mentioned above right?